File: 95-unique-binary-search-trees-ii/95-unique-binary-search-trees-ii.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def generateTrees(self, n):
        """
        :type n: int
        :rtype: List[TreeNode]
        """
        final = []
        def combine(temp,num,left,right):
            n1,n2 = len(left), len(right)
            for i in range(n1):
                for j in range(n2):
                    root = TreeNode(num)
                    root.left = left[i]
                    root.right = right[j]
                    temp.append(root)
        def generate(l):
            if not l:
                return [None]
            if len(l) == 1:
                return [TreeNode(l[0])]
            length = len(l)
            temp = []
            for i in range(length):
                left = generate(l[:i])
                right = generate(l[i+1:])
                combine(temp,l[i],left,right)
            return temp
        return generate([i for i in range(1,n+1)])
            
        
        

File: 95-unique-binary-search-trees-ii/95-unique-binary-search-trees-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<TreeNode> generateTrees(int n) {
        return generateSubTree(1, n);
    }
    public List<TreeNode> generateSubTree(int start, int end){
        List<TreeNode> res = new LinkedList<>();
        if(start > end){
            res.add(null);
            return res;
        }
        for(int i = start; i < end + 1; i++){
            List<TreeNode> leftTrees = generateSubTree(start, i-1);
            List<TreeNode> rightTrees = generateSubTree(i+1, end);
            for(TreeNode leftTree : leftTrees)
                for(TreeNode rightTree : rightTrees){
                    TreeNode root = new TreeNode(i);
                    root.left = leftTree;
                    root.right = rightTree;
                    res.add(root);
                }
        }
        return res;
    }
    

    
    
}